### PR TITLE
Supervision optimization

### PR DESCRIPTION
**Pull request by chenzhehuai**
Supervision optimization  to make the algorithm linear-time in the minibatch size, per minibatch, instead of quadratic-time.

- by:  change to call Concat(const Fst<Arc> &fst1, MutableFst<Arc> *fst2) 

- details:
```
previously we call "Concat(MutableFst<Arc> *fst1, const Fst<Arc> &fst2)" from AppendSupervision() by:

  for (int32 i = 0; i < input.size(); i++) {
    const Supervision &src = *(input[i]);
    fst::Concat(&output_supervision->back().fst, src.fst);
  }

now change to call "Concat(const Fst<Arc> &fst2, MutableFst<Arc> *fst1)" by

  for (int32 i =  input.size()-1; i >= 0; i--) {
    const Supervision &src = *(input[i]);
    fst::Concat(src.fst, &output_supervision->back().fst);
  }
```
___________

- run chain-supervision-test:
> /export/a12/zchen/works/concat/chain-supervision-test.log

- test whether the output of nnet3-chain-merge-egs is the same before and after modification:
```
zchen@g02:/export/a12/zchen/works/concat$  nnet3-chain-merge-egs --minibatch-size=256,128,64 ark:exp/chain/tdnn_lstm_1e_sp/egs/storage/2/cegs.1.ark ark,t:tmp/test3.txt
zchen@g02:/export/a12/zchen/works/concat$  nnet3-chain-merge-egs --minibatch-size=256,128,64 ark:exp/chain/tdnn_lstm_1e_sp/egs/storage/2/cegs.1.ark ark,t:tmp/test.txt
diff /export/a12/zchen/works/concat/tmp/test.txt /export/a12/zchen/works/concat/tmp/test3.txt
```

### CODE REVIEW

**Comment by chenzhehuai** [comment]
add things below
1. simplify the interface AppendSupervision by: i) delete compactify 2) use single output_supervision instead of vector
2. modify discriminative-supervision.cc correspondingly, along with test files  and other calling functions
3, test correctness:
```
zchen@g02:/export/a12/zchen/works/concat$ nnet3-chain-merge-egs --minibatch-size=256,128,64 ark:exp/chain/tdnn_lstm_1e_sp/egs/storage/2/cegs.1.ark ark,t:tmp/test4.txt
zchen@g02:/export/a12/zchen/works/concat$ diff tmp/test4.txt tmp/test3.txt     
```
4. test speedup:
```
/export/a12/zchen/works/concat/test_time.sh

zchen@login:/export/a12/zchen/works/concat$ grep real watanabe*
watanabe.log:real       3m28.864s
watanabe.log.new:real   0m23.637s
```

File: src/nnet3/discriminative-supervision.cc [hunk 1]
```diff
@@ -400,49 +400,36 @@ void DiscriminativeSupervisionSplitter::ComputeLatticeScores(const Lattice &lat,
...
-    (*output_supervision)[0] = *(input[0]);
+    (*output_supervision) = *(input[0]);
     return;
   }
-  std::vector<bool> output_was_merged;
-  output_supervision->clear();
-  output_supervision->reserve(input.size());
-  for (int32 i = 0; i < input.size(); i++) {
+  (*output_supervision) = *(input[num_inputs-1]);
+  for (int32 i = num_inputs-2; i > -1; i--) {
```
Comment (danpovey) [inline src/nnet3/discriminative-supervision.cc]: here and elsewhere, please put space around the operator -, i.e. 'num_inputs - 2'.  This is how we normally do it.  And I prefer 'i >= 0' to 'i > -1'.

File: src/nnet3/discriminative-supervision.cc [hunk 2]
```diff
@@ -400,49 +400,36 @@ void DiscriminativeSupervisionSplitter::ComputeLatticeScores(const Lattice &lat,
...
-  KALDI_ASSERT(output_was_merged.size() == output_supervision->size());
-  for (size_t i = 0; i < output_supervision->size(); i++) {
-    if (output_was_merged[i]) {
-      DiscriminativeSupervision &out_sup = (*output_supervision)[i];
-      fst::TopSort(&(out_sup.den_lat));
-      out_sup.Check();
+      KALDI_ERR << "mismatch between inputs";
     }
   }
+  DiscriminativeSupervision &out_sup = (*output_supervision);
```
Comment (danpovey) [inline src/nnet3/discriminative-supervision.cc]: no need for parentheses.

File: src/nnet3/discriminative-supervision.cc [hunk 3]
```diff
@@ -400,49 +400,36 @@ void DiscriminativeSupervisionSplitter::ComputeLatticeScores(const Lattice &lat,
...
-      output_was_merged.push_back(false);
-    }
-  }
-  KALDI_ASSERT(output_was_merged.size() == output_supervision->size());
-  for (size_t i = 0; i < output_supervision->size(); i++) {
-    if (output_was_merged[i]) {
-      DiscriminativeSupervision &out_sup = (*output_supervision)[i];
-      fst::TopSort(&(out_sup.den_lat));
-      out_sup.Check();
+      KALDI_ERR << "mismatch between inputs";
```
Comment (danpovey) [inline src/nnet3/discriminative-supervision.cc]: please make this a bit longer, and start with a capital.

**Comment by chenzhehuai** [comment]
done.
I also do a test on discriminative-supervision.cc, now it is the same

/export/a12/zchen/works/concat/test_disc.sh

**Comment by danpovey** [comment]
thanks!  Will merge when tests finish.

**Pull request merged**